File: Obfuscation/RedSiege/10_v1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Sample shellcode (in real scenarios this would be dynamically loaded)
unsigned char shellcode[] = "\xfc\x48\x83\xe4\xf0\xff\xd5";

// Function to pad the shellcode with NOPs (0x90)
unsigned char* pad_shellcode(unsigned char* shellcode, size_t size, size_t block_size, size_t* padded_size) {
    size_t pad_size = block_size - (size % block_size);
    if (pad_size == block_size) {
        *padded_size = size;
        return shellcode;  // No padding needed
    }
    
    *padded_size = size + pad_size;
    unsigned char* padded_shellcode = (unsigned char*)malloc(*padded_size);
    memcpy(padded_shellcode, shellcode, size);
    
    // Pad with NOPs
    memset(padded_shellcode + size, 0x90, pad_size);
    
    return padded_shellcode;
}

// Function to encode shellcode as MAC addresses
void encode_shellcode_as_mac(unsigned char* shellcode, size_t size) {
    printf("Encoded MAC Addresses:\n");
    
    for (size_t i = 0; i < size; i += 6) {
        // Print 6 bytes as a MAC address (dash-delimited)
        printf("%02x-%02x-%02x-%02x-%02x-%02x\n",
               shellcode[i], shellcode[i+1], shellcode[i+2],
               shellcode[i+3], shellcode[i+4], shellcode[i+5]);
    }
}

// Function to decode MAC addresses back into shellcode
void decode_mac_addresses(char* mac_addresses[], size_t num_macs, unsigned char* decoded_shellcode) {
    for (size_t i = 0; i < num_macs; i++) {
        unsigned int bytes[6];
        sscanf(mac_addresses[i], "%02x-%02x-%02x-%02x-%02x-%02x", 
               &bytes[0], &bytes[1], &bytes[2], 
               &bytes[3], &bytes[4], &bytes[5]);

        for (int j = 0; j < 6; j++) {
            decoded_shellcode[i * 6 + j] = (unsigned char)bytes[j];
        }
    }
}

int main() {
    size_t shellcode_size = sizeof(shellcode) - 1; // Exclude null terminator
    size_t padded_size;
    
    // Pad the shellcode to a multiple of 6 bytes
    unsigned char* padded_shellcode = pad_shellcode(shellcode, shellcode_size, 6, &padded_size);
    
    // Encode shellcode as MAC addresses
    encode_shellcode_as_mac(padded_shellcode, padded_size);
    
    // Example MAC addresses for decoding (would be dynamically generated in a real scenario)
    char* mac_addresses[] = {
        "fc-48-83-e4-f0-ff",
        "90-90-90-90-90-90"
    };
    
    // Allocate buffer to hold the decoded shellcode
    unsigned char* decoded_shellcode = (unsigned char*)malloc(padded_size);
    
    // Decode the MAC addresses back into shellcode
    decode_mac_addresses(mac_addresses, 2, decoded_shellcode);
    
    printf("\nDecoded Shellcode:\n");
    for (size_t i = 0; i < padded_size; i++) {
        printf("\\x%02x", decoded_shellcode[i]);
    }
    
    // Clean up allocated memory
    free(padded_shellcode);
    free(decoded_shellcode);

    return 0;
}

File: Obfuscation/RedSiege/10_v1.py
import re

# Sample shellcode
shellcode = b"\xfc\x48\x83\xe4\xf0\xff\xd5"

# Helper function to pad shellcode if it's not a multiple of 6 bytes
def pad_shellcode(shellcode, block_size=6):
    pad_size = block_size - (len(shellcode) % block_size)
    return shellcode + b'\x90' * pad_size if pad_size < block_size else shellcode

# Function to encode shellcode into MAC addresses
def encode_shellcode_as_mac(shellcode):
    shellcode = pad_shellcode(shellcode)  # Pad if necessary
    mac_addresses = []
    
    # Process the shellcode in 6-byte chunks (MAC address length)
    for i in range(0, len(shellcode), 6):
        chunk = shellcode[i:i + 6]
        mac_address = '-'.join(f"{byte:02x}" for byte in chunk)
        mac_addresses.append(mac_address)
    
    return mac_addresses

# Function to decode MAC addresses back to shellcode
def decode_mac_addresses(mac_addresses):
    decoded_shellcode = b""
    
    for mac in mac_addresses:
        # Remove the dash and convert hex back to bytes
        mac_bytes = bytes.fromhex(mac.replace('-', ''))
        decoded_shellcode += mac_bytes
    
    return decoded_shellcode

# Example usage:
mac_addresses = encode_shellcode_as_mac(shellcode)
print("Encoded MAC Addresses:")
for mac in mac_addresses:
    print(mac)

decoded_shellcode = decode_mac_addresses(mac_addresses)
print("\nDecoded Shellcode:")
print(decoded_shellcode)

# Check if the decoded shellcode matches the original
if decoded_shellcode == pad_shellcode(shellcode):
    print("\nDecoding was successful and matches the original shellcode.")
else:
    print("\nDecoding failed or the shellcode doesn't match.")

